### PR TITLE
Themes: Cache renderToString output & add stats

### PR DESCRIPTION
**Pull request by ehg**
Try memoizing the costly renderToString call.

### CODE REVIEW

**Comment by ehg** [comment]
A few server-reported response times. It seems there's not much difference in the development environment.

`renderToString`, no caching:

screen shot 2016-01-18 at 12 44 05


`renderToString`, lodash `memoize` ( this PR ):
screen shot 2016-01-18 at 12 41 56


`renderToString`, cache markup in var:
screen shot 2016-01-18 at 12 38 34

**Comment by seear** [comment]
These techniques should prove very effective with more complex component trees, so it is tempting to deploy one of them. However, I think we should keep them in the bag until we can show some real benefit, because until then, caching would only serve as a potential source of bugs.

**Comment by seear** [comment]
> I think we should keep them in the bag until we can show some real benefit, because until then, caching would only serve as a potential source of bugs.

I've changed my mind on this. Our big concern it that `renderToString` may block the event loop, and we don't know of a good way to find out if this is the case. So, if we turn this caching on and show that we get mostly cache hits, then we know that `renderToString` is being called so infrequently as to be nothing to worry about.

**Comment by mcsf** [comment]
I welcome any sort of experimenting with cache. My one concern here is with memoization — specifically, relying on implementations like lodash's — due to memory management within the memoization closure.

AFAIK, React factory calls are not pure (are they?), which means that we have no guarantees regarding what gets passed to `memoizedRender` (in order to answer: how do we know we have a cached output for that input?). So, memoization should 1) know how to compare React trees, _or_ 2) be moved higher up, caring only about the relevant input of each request (thus avoiding calling React at all); aside, it should 3) manage memory internally, possibly with a round-robin structure limited to _n_ slots. 

On the other hand, the `cachedLayout`-based solution seems a lot easier to get _right_ as an MVP. 

Summing up: a more sophisticated solution based on function memoization could be more interesting — and exportable to other parts of SSR'd Calypso — but needs more conceptual work and testing.

**Comment by ehg** [comment]
> AFAIK, React factory calls are not pure (are they?), which means that we have no guarantees regarding what gets passed to memoizedRender (in order to answer: how do we know we have a cached output for that input?)

Interesting :)

I'm not asserting that the factory call is pure, but I noticed that calls to the factory always produce cache hits, presumably because the string representation is always the same on the same node instance.

I agree that `cachedLayout` would be a safer MVP. Once we get to the stage of wanting to render dynamic data, we'll have to look at more distributed solutions, e.g. memcache or a CDN.

**Comment by mcsf** [comment]
[Tested](http://codepen.io/mcsf/pen/WrXyde) in Codepen:

```js
const allEq = (xs) =>
  xs.reduce((cur, next) => ! next || cur === next)

const test = _.compose(x => x.toString(), allEq)

const Component = () => <h1>Woot!</h1>

const Factory = React.createFactory(Component)

const cs = [ <Component />, <Component /> ]
const fs = [ Factory(), Factory() ]
const ss = fs.map(React.renderToString)

React.render(<ul>
  <ol>dummy test: { test([ 4, 4 ]) }</ol>
  <ol>components test: { test(cs) }</ol>
  <ol>factories output test: { test(fs) }</ol>
  <ol>rendered strings test: { test(ss) } // ultimately different because of React ids</ol>
</ul>, document.body)
```

yields

> dummy test: true
> components test: false
> factories output test: false
> rendered strings test: false // ultimately different because of React ids

**Comment by ehg** [comment]
> Tested in Codepen:

Nice :) So they're not the same objects, but they *do* consistently output the same string, e.g. test the `JSON.stringify()`ied output is equal.

**Comment by ehg** [comment]
I switched to a simple var cache, as opposed to a `memoize` in 392d58c, in addition to some basic stats.

**Comment by ehg** [comment]
Pinging @seear to check the stats stuff :)

File: server/pages/index.js [hunk 1]
```diff
@@ -377,7 +386,19 @@ module.exports = function() {
...
+					if ( cachedDesignMarkup ) {
+						context.layout = cachedDesignMarkup;
+					} else {
+						let startTime = Date.now();
+						context.layout = cachedDesignMarkup = ReactDomServer.renderToString( LayoutLoggedOutDesignElement );
+						let rtsTimeMs = Date.now() - startTime;
+
+						if ( rtsTimeMs > 15 ) {
+							// We think that renderToString should generally never take more than 15ms. We're probably wrong.
+							bumpStat( 'calypso-ssr', 'loggedout-design-over-15ms-rendertostring' );
```
Comment (ehg) [inline server/pages/index.js]: We probably only want to bump stats in production.

**Comment by seear** [comment]
Maybe restrict the stat bump to prod?

File: server/pages/index.js [hunk 2]
```diff
@@ -314,6 +316,13 @@ function renderLoggedInRoute( req, res ) {
 	}
 }
 
+function bumpStat( group, name ) {
+	const url = `http://pixel.wp.com/g.gif?v=wpcom-no-pv&x_${ group }=${ name }&t=${ Math.random() }`;
+	console.log( 'bumping', group, name, url );
```
Comment (seear) [inline server/pages/index.js]: don't forget to take this line out before merge

**Comment by seear** [comment]
Looks sensible to me. Maybe in time we will want to add some extra tiers for the time stats.
:+1:

**Comment by mcsf** [comment]
Just so we can find it later, here's a [codepen seeking to measure React tree construction](http://codepen.io/mcsf/pen/VerGZL?editors=001) using a `connect`-like wrapping component. (If someone decides to play with it, note the discrepancies between the measured durations and the actual time the program takes to render.)
